### ISSUE TITLE
Add support for linking Scala code as a native library

### ISSUE THREAD

**Issue by densh** [scala-native/scala-native#155]
We already have an interop layer across C ABI: `@extern object`. 

- [ ] Add support for linking without injecting `main` entry point in the compiler
- [ ] Don't disallow having non-extern methods in extern objects (https://github.com/scala-native/scala-native/issues/79)

This will let our toolchain produce "*.so"/"*.dylib"/"*.dll" files.

**Comment by RimashMohomed** [comment]
@densh interesting to contribute for this feature can you please guide me through

**Comment by densh** [comment]
We need to cook this one a bit longer before we are confident to include it in the official feature set.

**Comment by drdozer** [comment]
I'm potentially interested in this. I can imagine aggressively compiling kernels that don't need to allocate objects down to native code that's basically just maths and pointers, and then loading this back in as a black-box native routine. I'm thinking things like lookups on static tables and indexes.

**Comment by densh** [comment]
Related: #450

**Comment by velvia** [comment]
This would be an awesome feature.  How cooked is the SBT setting?

**Comment by densh** [comment]
@velvia Roughly as raw as meat in steak tartare.

**Comment by nornagon** [comment]
It'd be super neat to be able to dynamically load SN-generated `*.so` files from SN, and call functions in them. I'm imagining a live-reload scenario for game dev :D

**Comment by jstoiko** [comment]
How big would the resulting dynamic library be? Is there any way to get an idea?

**Comment by densh** [comment]
@jstoiko Our current output for the applications is in low single-digit megabytes. We haven't spent much time optimizing for binary size yet. At the moment our goal is to optimize for runtime performance first and foremost.

**Comment by rwhaling** [comment]
.so files could also be useful for integration with Nginx Unit, which dynamically loads language-specific modules with dlopen().

**Comment by hepin1989** [comment]
Highly needed, my friend @WayneWang12 can't use scala-native for the competition of alicloud...

**Comment by hepin1989** [comment]
Any update on this? Would be very interesting to implement out tql with this.

**Comment by Zizico2** [comment]
Any progress?

**Comment by sheiksadique** [comment]
I see that there is an open PR. Are there plans to continue this work and merge the PR?

**Comment by WojciechMazur** [comment]
> I see that there is an open PR. Are there plans to continue this work and merge the PR?

@sheiksadique 
Yes, we are going to continue work in this area. It definitely still needs proper review and a good set of additional tests (especially in terms of performance and ABI compatibility)

**Comment by sheiksadique** [comment]
In the mean time, I wanted to to try out this PR but ran into a wall (I guess ignorance is the wall I need to scale). 

https://stackoverflow.com/questions/68161110/sbt-scala-plugin-from-github-pr-scalanative-fails-with-value-not-found

Could you point me to how I could use the PR branch in my current scala-native project?

**Comment by letalvoj** [comment]
Seems that https://github.com/scala-native/scala-native/pull/2145 is set for 0.5 milestone. Nice.

**Comment by armanbilge** [comment]
I added an update about the status of this feature in https://github.com/scala-native/scala-native/pull/2145#issuecomment-1174575616.
